### PR TITLE
    Makes `OverlayDialogFactory` more consistent with `ScreenViewFactory`.

### LINKED ISSUE #713: Polish up OverlayDialogFactory, esp. Modality
Should change `buildDialogForView` to accept `ScreenViewHolder` instead.

Probably should update `OverlayDialogFactory` to work more like `ScreenViewFactory` -- make `DialogHolder` public, and eliminate the separate `update` method.

_Originally posted by @rjrjr in https://github.com/square/workflow-kotlin/pull/712#discussion_r838792858_

### PR DESCRIPTION
**Pull request by rjrjr**
`OverlayDialogFactory` now has one method that produces `OverlayDialogHolder` wrappers, just like `ScreenViewFactory` and `ScreenViewHolder`.

I'm sad about all the copy/paste API here, but every time I try to consolidate things into a single set of `UiFactory` / `UiHolder` / `UiRunner` interfaces, I wind up with an opaque mess of parameter types that usually breaks things. A `Dialog` is not a `View`, both classes are dying anyway one hopes, and my attempts to factor out a pretty shared wrapper API for them has been a really bad use of time.

Also note:

- Implementors of `ScreenOverlayDialogFactory` now have nearly complete control over the `ScreenViewFactory` that builds the content view.

- `ModalScreenOverlayOnBackPressed` is now `ModalScreenOverlayBackButtonHelper`, and eliminates all direct dependencies on Jetpack's  `OnBackPressedDispatcherOwner` -- necessary for we poor souls who remain incompatible with that.

- We now take care only to apply the back button work for modals. We sure don't want toasts intercepting the back button, or anything else.

The first commit avoids renaming anything interesting, for better diffs and ease of review.

Replaces #804, closes #713.

### CODE REVIEW

File: workflow-ui/core-android/src/main/java/com/squareup/workflow1/ui/container/ScreenOverlayDialogFactory.kt [hunk 1]
```diff
@@ -29,29 +28,101 @@ import kotlin.reflect.KClass
...
+ * [back press handlers][com.squareup.workflow1.ui.backPressedHandler] on covered
+ * views and windows should not fire. [ScreenOverlayDialogFactory] takes care
+ * of that requirement by default, presuming that your app uses Jetpack
+ * [OnBackPressedDispatcher][androidx.activity.OnBackPressedDispatcher]. If that is not
+ * the case, alternative handling can be provided via [ModalScreenOverlayBackButtonHelper].
+ *
+ * It is important to note that the modal behavior described here is all is keyed to the
+ * [ModalOverlay] interface, not its parent type [Overlay]. Rendering types that declare the
+ * latter but not the former can be used to create dialogs for non-modal windows like toasts
+ * and tool tips.
```
Comment (steve-the-edwards) [inline workflow-ui/core-android/src/main/java/com/squareup/workflow1/ui/container/ScreenOverlayDialogFactory.kt:96]: 👨🏻‍🍳 💋 . Thanks for this. Very helpful!

File: workflow-ui/core-android/src/main/java/com/squareup/workflow1/ui/container/DialogSession.kt [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+import com.squareup.workflow1.ui.WorkflowUiExperimentalApi
+import com.squareup.workflow1.ui.androidx.WorkflowLifecycleOwner
+import com.squareup.workflow1.ui.androidx.WorkflowSavedStateRegistryAggregator
+
+/**
+ * Used by [LayeredDialogSessions] to manage lifecycle and view persistence concerns for an
+ * [OverlayDialogHolder], as well as enforcing modal behavior.
+ */
+@WorkflowUiExperimentalApi
+internal class DialogSession(
```
Comment (steve-the-edwards) [inline workflow-ui/core-android/src/main/java/com/squareup/workflow1/ui/container/DialogSession.kt:24]: What I found I wanted on these Session classes is a definition of the lifecycle of a session. I know its internal and an implementation detail - but I still would love to know :).

File: workflow-ui/core-android/src/main/java/com/squareup/workflow1/ui/container/ScreenOverlayDialogFactory.kt (hunk 1 above)
Comment (steve-the-edwards) [reply workflow-ui/core-android/src/main/java/com/squareup/workflow1/ui/container/ScreenOverlayDialogFactory.kt:96]: Btw. Is this linked from `BodyAndModalsContainer` kdoc?

**Review by steve-the-edwards — APPROVED** [review]

**Comment by rjrjr** [comment]
@steve-the-edwards I just pushed a new commit with a lot more kdoc, PTAL.

File: workflow-ui/core-android/src/main/java/com/squareup/workflow1/ui/container/OverlayDialogHolder.kt [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+  public val dialog: Dialog
+
+  /** The [ViewEnvironment] that was provided the last time [dialog] was updated by [runner]. */
+  public val environment: ViewEnvironment
+
+  /**
+   * The function that is run by [show] to update [dialog] with a new [Screen] rendering and
+   * [ViewEnvironment].
+   */
+  public val runner: (rendering: OverlayT, environment: ViewEnvironment) -> Unit
```
Comment (helios175) [inline workflow-ui/core-android/src/main/java/com/squareup/workflow1/ui/container/OverlayDialogHolder.kt:30]: I thought it'd be the runner factory at first glance. Lol.
Why not update?
(no biggies)

**Comment by rjrjr** [comment]
Internal Square tests are all green with this, so I'll def. publish immediately after merging.

In the meantime, I've published `1.8.0-beta02-SNAPSHOT`.

File: workflow-ui/core-android/src/main/java/com/squareup/workflow1/ui/container/RealOverlayDialogHolder.kt [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+  override val dialog: Dialog,
+  runner: (rendering: OverlayT, environment: ViewEnvironment) -> Unit
+) : OverlayDialogHolder<OverlayT> {
+
+  private var _environment: ViewEnvironment = initialEnvironment
+  override val environment: ViewEnvironment get() = _environment
+
+  override val runner = { newScreen: OverlayT, newEnvironment: ViewEnvironment ->
+    _environment = newEnvironment
+    runner(newScreen, newEnvironment)
```
Comment (vgonda) [inline workflow-ui/core-android/src/main/java/com/squareup/workflow1/ui/container/RealOverlayDialogHolder.kt:19]: Is there worthwhile naming that would make it more clear that this `runner` lambda declared on line 19 is calling the `runner` lambda on line 11? Took a second to realize there were two different lambdas named `runner` and we're not intentionally causing a stack overflow lol

**Review by vgonda — COMMENTED** [review]
Thanks for the kdoc! I'm not super familiar with all this, so it was really helpful for me.

Nothing blocking on my end!

File: workflow-ui/core-android/src/main/java/com/squareup/workflow1/ui/container/BodyAndOverlaysContainer.kt [hunk 5]
```diff
@@ -19,6 +19,16 @@ import com.squareup.workflow1.ui.ViewEnvironment
 import com.squareup.workflow1.ui.WorkflowUiExperimentalApi
 import com.squareup.workflow1.ui.WorkflowViewStub
 
+/**
+ * Default container for [Overlay] renderings, providing support for
+ * specializing subtypes [ModalOverlay] and [ScreenOverlay]. As much
```
Comment (steve-the-edwards) [inline workflow-ui/core-android/src/main/java/com/squareup/workflow1/ui/container/BodyAndOverlaysContainer.kt:24]: orthogonal subtypes...

File: workflow-ui/core-android/src/main/java/com/squareup/workflow1/ui/container/BodyAndOverlaysContainer.kt [hunk 6]
```diff
@@ -19,6 +19,16 @@ import com.squareup.workflow1.ui.ViewEnvironment
...
+ * work as possible is delegated to the public [LayeredDialogSessions]
+ * support class, to make it practical to write custom forks should
+ * the need arise.
+ *
+ * See [ScreenOverlayDialogFactory] for an general overview of how
```
Comment (steve-the-edwards) [inline workflow-ui/core-android/src/main/java/com/squareup/workflow1/ui/container/BodyAndOverlaysContainer.kt:29]: nit: 
```suggestion
 * See [ScreenOverlayDialogFactory] for a general overview of how
```

File: workflow-ui/core-android/src/main/java/com/squareup/workflow1/ui/container/LayeredDialogSessions.kt [hunk 7]
```diff
@@ -17,15 +17,57 @@ import com.squareup.workflow1.ui.WorkflowUiExperimentalApi
...
+ * - And the Dialog is [shown][android.app.Dialog.show]
+ *
+ * The [DialogSession] is maintained (and its two flavors of View state are recorded
+ * as prompted by the Android runtime) so long as [update] is called with a
+ * [Compatible] [Overlay] rendering in the same position.
+ *
+ * When [update] is called without a matching [Overlay], or the
+ * [parent Lifecycle][getParentLifecycleOwner] ends, the [DialogSession] ends,
+ * its [ViewTreeLifecycleOwner] is destroyed, and the Dialog is
+ * [dismissed][android.app.Dialog.dismiss].
```
Comment (steve-the-edwards) [inline workflow-ui/core-android/src/main/java/com/squareup/workflow1/ui/container/LayeredDialogSessions.kt:70]: Ok, now I get it. This is a great summary. Thanks.

File: workflow-ui/core-android/src/main/java/com/squareup/workflow1/ui/container/RealOverlayDialogHolder.kt (hunk 4 above)
Comment (steve-the-edwards) [reply workflow-ui/core-android/src/main/java/com/squareup/workflow1/ui/container/RealOverlayDialogHolder.kt:19]: `realRunner` lol.

**Review by steve-the-edwards — APPROVED** [review]

File: workflow-ui/core-android/src/main/java/com/squareup/workflow1/ui/container/AndroidOverlay.kt [hunk 8]
```diff
@@ -6,6 +6,15 @@ import com.squareup.workflow1.ui.WorkflowUiExperimentalApi
...
  *
+ * Note that you can mix this interface with others:
+ *
+ * - [ModalOverlay] to indicate that user input to covered views and dialogs
+ *   should be blocked while this one is visible.
+ *
+ * - [ScreenOverlay] for dialogs whose content is defined by a wrapped
+ *   [Screen][com.squareup.workflow1.ui.Screen] instance. And in this case,
+ *   also note [ScreenOverlayDialogFactory].
+ *
```
[the remark this answers is not in the document]
Comment (rjrjr) [reply workflow-ui/core-android/src/main/java/com/squareup/workflow1/ui/container/AndroidOverlay.kt:17]: Yeah, that's up to the library. There's a sample that demonstrates one way to get that done:

https://github.com/square/workflow-kotlin/blob/main/samples/containers/common/src/main/java/com/squareup/sample/container/panel/ScrimScreen.kt

https://github.com/square/workflow-kotlin/blob/main/samples/containers/android/src/main/java/com/squareup/sample/container/ScrimContainer.kt

Applied here:

https://github.com/square/workflow-kotlin/blob/main/samples/tictactoe/common/src/main/java/com/squareup/sample/mainworkflow/TicTacToeWorkflow.kt#L103

File: workflow-ui/core-android/src/main/java/com/squareup/workflow1/ui/container/RealOverlayDialogHolder.kt (hunk 4 above)
Comment (rjrjr) [reply workflow-ui/core-android/src/main/java/com/squareup/workflow1/ui/container/RealOverlayDialogHolder.kt:19]: Oh, that's not good. Will tidy.

File: workflow-ui/core-android/src/main/java/com/squareup/workflow1/ui/container/AndroidOverlay.kt [hunk 9]
```diff
@@ -6,6 +6,15 @@ import com.squareup.workflow1.ui.WorkflowUiExperimentalApi
  * Interface implemented by a rendering class to allow it to drive an Android UI
  * via an appropriate [OverlayDialogFactory] implementation.
  *
+ * Note that you can mix this interface with others:
+ *
+ * - [ModalOverlay] to indicate that user input to covered views and dialogs
+ *   should be blocked while this one is visible.
```
[the remark this answers is not in the document]
Comment (rjrjr) [reply workflow-ui/core-android/src/main/java/com/squareup/workflow1/ui/container/AndroidOverlay.kt:12]: So relieved you saw that unprompted.

File: workflow-ui/core-android/src/main/java/com/squareup/workflow1/ui/container/AlertOverlayDialogFactory.kt [hunk 10]
```diff
@@ -53,29 +52,26 @@ public open class AlertOverlayDialogFactory : OverlayDialogFactory<AlertOverlay>
...
-    environment: ViewEnvironment
-  ) {
-    (dialog as AlertDialog).apply {
-      if (rendering.cancelable) {
-        setOnCancelListener { rendering.onEvent(Canceled) }
-        setCancelable(true)
-      } else {
-        setCancelable(false)
-      }
+      .let { alertDialog ->
```
[the remark this answers is not in the document]
Comment (rjrjr) [reply workflow-ui/core-android/src/main/java/com/squareup/workflow1/ui/container/AlertOverlayDialogFactory.kt:55]: Usually I merge them, sloppy.

**Pull request merged**